src/tasks/submit/flashbots/provider.rs

```diff
@@ -0,0 +1,109 @@
+//! A generic Flashbots bundle API wrapper.
+use crate::config::BuilderConfig;
+use alloy::{
+    primitives::{BlockNumber, keccak256},
+    rpc::types::mev::{EthBundleHash, MevSendBundle, SimBundleResponse},
+    signers::Signer,
+};
+use eyre::Context as _;
+use init4_bin_base::utils::signer::LocalOrAws;
+use reqwest::header::CONTENT_TYPE;
+use serde_json::json;
+
+/// A wrapper over a `Provider` that adds Flashbots MEV bundle helpers.
+#[derive(Debug)]
+pub struct Flashbots {
+    /// The base URL for the Flashbots API.
+    pub relay_url: url::Url,
+    /// Builder configuration for the task.
+    pub config: BuilderConfig,
+    /// Signer is loaded once at startup.
+    signer: LocalOrAws,
+}
+
+impl Flashbots {
+    /// Wraps a provider with the URL and returns a new `FlashbotsProvider`.
+    pub async fn new(config: &BuilderConfig) -> Self {
+        let relay_url =
+            config.flashbots_endpoint.as_ref().expect("Flashbots endpoint must be set").clone();
+
+        let signer =
+            config.connect_builder_signer().await.expect("Local or AWS signer must be set");
+
+        Self { relay_url, config: config.clone(), signer }
+    }
+
+    /// Sends a bundle  via `mev_sendBundle`.
+    pub async fn send_bundle(&self, bundle: MevSendBundle) -> eyre::Result<EthBundleHash> {
+        let params = serde_json::to_value(bundle)?;
+        let v = self.raw_call("mev_sendBundle", params).await?;
+        let hash: EthBundleHash =
+            serde_json::from_value(v.get("result").cloned().unwrap_or(serde_json::Value::Null))?;
+        Ok(hash)
+    }
+
+    /// Simulate a bundle via `mev_simBundle`.
+    pub async fn simulate_bundle(&self, bundle: MevSendBundle) -> eyre::Result<()> {
+        let params = serde_json::to_value(bundle)?;
+        let v = self.raw_call("mev_simBundle", params).await?;
+        let resp: SimBundleResponse =
+            serde_json::from_value(v.get("result").cloned().unwrap_or(serde_json::Value::Null))?;
+        dbg!("successfully simulated bundle", &resp);
+        Ok(())
+    }
+
+    /// Fetches the bundle status by hash
+    pub async fn bundle_status(
+        &self,
+        _hash: EthBundleHash,
+        block_number: BlockNumber,
+    ) -> eyre::Result<()> {
+        let params = json!({ "bundleHash": _hash, "blockNumber": block_number });
+        let _ = self.raw_call("flashbots_getBundleStatsV2", params).await?;
+        Ok(())
+    }
+
+    /// Makes a raw JSON-RPC call with the Flashbots signature header to the method with the given params.
+    async fn raw_call(
```

Comment:
Alloy itself has a method for performing raw calls: https://github.com/alloy-rs/alloy/blob/main/crates/provider/src/provider/trait.rs#L1182-L1213

We could instead make a `FlashbotsExt: Provider` trait that has these fns and implement that for the alloy provider we use. I think this would be a tad cleaner

Reply:
alloy odesn't allow passing the necessary header, unfortunately, so there needs to be some sort of duplication here